enhancements/compound-bundle-constraints.md

```diff
@@ -0,0 +1,350 @@
+
+However, what if:
+- `foo.v1.0.0` can use multiple versions of the same API, but only one of that group-kind?
+- `foo.v1.0.0` can use multiple versions of the same API, but only if another package version is present?
+- `foo.v1.0.0` cannot use a particular API version, but can use all others?
+
+These constraints are logical disjunctions and negations in character; the existing
+top-level constraints are together a logical conjunction, although not explicit.
+Currently the spec does not support specifying such "compound" dependency constraints.
+
+### Goals
+
+- Define logical conjunction, disjunction, and negation properties as dependency constraints, termed "compound constraints".
+- Define nesting of compound constraints constraints.
+- Describe how compound constraints are evaluated.
+- Describe what needs to change in OLM's resolver to support compound constraints.
+- Plan for implementation under an alpha feature gate, and a timeline for GA.
+
+### Non-Goals
+
+- Change the existing constraint definitions.
+- Break bundle spec or OLM backwards-compatibility in any way.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator author, I should be able to specify "any of these APIs"
+or "none of these APIs" dependencies as bundle properties, and "all of these APIs" explicitly.
+
+#### Story 2
+
+As an operator author, I should be able to nest any compound constraint(s)
+within a parent compound constraint.
+
+#### Story 3
+
+As a cluster administrator using OLM, I should receive the same information
+about successful and failed resolution as with current "simple" constraints.
+
+#### Story 4
+
+As a cluster administrator, I should be able to opt-in to resolving compound
+constraints in OLM while the feature is alpha. When opted-out, the resolver
+should emit errors so bundles are not installed incorrectly, ex. missing an API.
+
+### Implementation Details/Notes/Constraints
+
+**Note**: see ["Implementation History"](#implementation-history) for proof-of-concept info.
+
+I propose the new properties `olm.all.required`, `olm.any.required`, and `olm.none.required`
+to specify conjunction, disjunction, and negation compound constraints, respectively.
+Their value type is a list of `olm.package.required`, `olm.gvk.required`, or any
+compound constraint listed above.
+
+#### Conjunction and disjunction
+
+These compound constraint types are evaluated following their logical definitions.
+
+This is an example of a conjunctive constraint of two packages and one GVK,
+i.e. they must all be satisfied by installed bundles:
+
+```yaml
+schema: olm.bundle
+name: baz.v1.0.0
+properties:
+- type: olm.all.required
+  value:
+  - type: olm.package.required
+    value:
+      packageName: baz
+      versionRange: '>=1.0.0'
+  - type: olm.package.required
+    value:
+      packageName: buf
+      versionRange: '>=1.0.0'
+  - type: olm.gvk.required
+    value:
+      group: bazs.example.com
+      version: v1
+      kind: Baz
```

Comment:
Should `value` be a struct with a string `description` and a slice of sub-`constraints`. I seem to recall some prototype code looking like that.

Reply:
A description would be useful! I am in favor of changing `value` like that.